src/protocol/boolean/bit_decomposition.rs

```diff
@@ -38,49 +40,170 @@ impl BitDecomposition {
         S: LinearSecretSharing<F> + BasicProtocols<C, F>,
         C: Context + RandomBits<F, Share = S>,
     {
-        // step 1 in the paper is just describing the input, `[a]_p` where `a ∈ F_p`
-
-        // Step 2. Generate random bitwise shares
+        // Step 1. Generate random bitwise shares [r]_B and linear share [r]_p
         let r = rbg.generate(record_id).await?;
 
-        // Step 3, 4. Reveal c = [a - b]_p
-        let c = (a_p.clone() - &r.b_p)
+        // Step 2: Reveal c = [a - r]_p
+        let c: F = (a_p.clone() - &r.b_p)
             .reveal(ctx.narrow(&Step::RevealAMinusB), record_id)
             .await?;
 
-        // Step 5. Add back [b] bitwise. [d]_B = BitwiseSum(c, [b]_B) where d ∈ Z
-        //
-        // `BitwiseSum` outputs one more bit than its input, so [d]_B is (el + 1)-bit long.
-        let d_b = add_constant(ctx.narrow(&Step::AddBtoC), record_id, &r.b_b, c.as_u128()).await?;
+        // Step 2.1: Edge case, if r is coincidentally a bit decomposition of [a]_p, we're done.
+        if c == F::ZERO {
+            return Ok(r.b_b);
+        }
 
-        // Step 6. q = d >=? p (note: the paper uses p <? d, which is incorrect)
-        let q_p = BitwiseLessThanPrime::greater_than_or_equal_to_prime(
-            ctx.narrow(&Step::IsPLessThanD),
+        // Step 3. Compute [q]_p = 1 - [r <_B p - c ]_p. q is 1 iff r + c >= p in the integers.
+        let p_minus_c: u128 = F::PRIME.into() - c.as_u128();
+        let q_p = bitwise_less_than_constant(
+            ctx.narrow(&Step::IsRLessThanPMinusC),
             record_id,
-            &d_b,
+            &r.b_b,
+            p_minus_c.into(),
         )
         .await?;
 
-        // Step 7. a bitwise scalar value `f_B = bits(2^el - p)`
+        // Step 4 has a lot going on. We'd going to break it down into substeps.
+        // Step 4.1. Make a bitwise scalar value of f = 2^el + c - p.
+        // let el = usize::try_from(u128::BITS - F::PRIME.into().leading_zeros()).unwrap();
         let el = u128::BITS - F::PRIME.into().leading_zeros();
```

Comment:
is it just `let el = F::BITS`?

Reply:
Maybe? I was just modifying what's used in [add_constant.rs](https://github.com/private-attribution/ipa/blob/main/src/protocol/boolean/add_constant.rs#L144). Looking at the blame, it seems like @martinthomson did this to remove the trailing bits.

Reply:
ah no, I think this version is better if PRIME << $2^n-1$